### PR TITLE
My fixes

### PR DESCRIPTION
**Pull request by ppodolsky**
I've led experiments to working state (but I have not done all we need, of course). Also, I fixed profile controller, added little i18n, remade is_role? in User model. There were several columns I added (they are in last migration now)

### CODE REVIEW

File: app/controllers/profiles_controller.rb [hunk 1]
```diff
@@ -1,30 +1,46 @@
...
         :secondary_email,
         :first_name,
         :last_name,
         :phone,
         :gender,
         :ethnicity,
-        :age,
+        :birth_year,
         :class_year,
         :total_years,
```
Comment (mlfreer) [inline app/controllers/profiles_controller.rb]: may be now we can remove it? cause we have year_started?

Comment (ppodolsky) [inline app/controllers/profiles_controller.rb]: Yes, we can:) I'll do it

**Comment by jrhorn424** [comment]
Well, now I feel foolish. I'm reviewing now.

@PaulPodolsky Which tickets can be closed when this merge? You can just reply with the hash number, like #177 to close that ticket.

**Comment by ppodolsky** [comment]
#20 #13 #14 #15 #17

**Comment by mlfreer** [comment]
also in further pull requests please wright in the title which tickets it is on, so we'll be able easily recognize and review it

File: Gemfile [hunk 2]
```diff
@@ -68,6 +68,9 @@ gem 'bootstrap-generators', github: 'decioferreira/bootstrap-generators'
 # http://devise.plataformatec.com.br/
 gem 'devise', '~> 3.1.0' # authentication
 
+#https://github.com/CanCanCommunity/cancancan
+gem 'cancancan', '~> 1.7'
+
```
Comment (jrhorn424) [inline Gemfile]: This is ok, but I'm more familiar with authority. If you want to go this route, let's remove authority.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -9,8 +9,8 @@ class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
 
 
-  def after_sign_in_path_for(resource)
-    return dashboard_path if resource.is_admin?
-    return user_profile_path(resource)
+  def after_sign_in_path_for(user)
+    return dashboard_path if user.is_administrator?
+    return profile_path
```
Comment (jrhorn424) [inline app/controllers/application_controller.rb:14]: Do we need to pass in `resource`? That's something expected by Devise (and Inherited Resources), but I'm not sure if having a singleton resource takes care of that. I've not used them before.

File: app/assets/javascripts/experiments.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+jQuery(document).ready(function($) {
+    $("[data-href]").click(function() {
+        window.document.location = $(this).data("href");
+    });
+    $(".remove-link").on("ajax:success", function(e, data, status, xhr){
+        alert("The post was deleted.");
+    });
+});
```
Comment (jrhorn424) [inline app/assets/javascripts/experiments.js:8]: What's the context around this addition, @PaulPodolsky? I thought similar functionality was already provided by the rails unobtrusive javascript.

File: app/controllers/experiments_controller.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
+class ExperimentsController < InheritedResources::Base
+  actions :new,:create,:index,:show,:update,:destroy
+
+  def index
+    @experiments = Experiment.all
+  end
```
Comment (jrhorn424) [inline app/controllers/experiments_controller.rb]: This isn't necessary.

File: app/controllers/experiments_controller.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
+class ExperimentsController < InheritedResources::Base
```
Comment (jrhorn424) [inline app/controllers/experiments_controller.rb:1]: `InheritedResources::Base` gives us basic crud actions for free. If you need to override it, just write the processing you need before or after a call to super for the action using the action name.

For example, if you want to downcase all experiment names, you could do:

``` ruby
...
def index
  @experiments = Experiment.all
  @experiments.map! { |e| e.downcase }
  index!
end
...
```

File: app/controllers/experiments_controller.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  actions :new,:create,:index,:show,:update,:destroy
+
+  def index
+    @experiments = Experiment.all
+  end
+  def new
```
Comment (jrhorn424) [inline app/controllers/experiments_controller.rb]: Not needed.

File: app/controllers/experiments_controller.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    @experiment = Experiment.new
+  end
+  def show
+    @experiment = Experiment.find(params[:id])
```
Comment (jrhorn424) [inline app/controllers/experiments_controller.rb]: Not needed.

File: Gemfile (hunk 2 above)
Comment (ppodolsky) [inline Gemfile]: We will discuss it with Misha tomorrow. Should understand if we need authorities.

File: app/controllers/experiments_controller.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    @sessions = @experiment.sessions
```
Comment (jrhorn424) [inline app/controllers/experiments_controller.rb]: This would work if one did

``` ruby
@sessions = @experiment.sessions
show!
```

File: app/assets/javascripts/experiments.js (hunk 4 above)
Comment (ppodolsky) [inline app/assets/javascripts/experiments.js:8]: TR couldn't be wrapped by A tag. The second one (.remove-link) is nesessary for removing row by js, without reloading the entire page. Dunno if rails provides some shortcuts for this.

File: app/controllers/experiments_controller.rb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    @experiment = Experiment.find(params[:id])
+    @experiment.destroy
+    respond_to do |format|
+      format.html { redirect_to experiments_path }
+      format.js   { render :layout=>false }
+    end
+    @experiments = Experiment.all
+  end
+  private
+  def experiment_params
```
Comment (jrhorn424) [inline app/controllers/experiments_controller.rb]: If you follow the pattern in [InheritedResources](https://github.com/josevalim/inherited_resources), actions will have access to the appropriate params defined here.

File: app/controllers/experiments_controller.rb [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    @experiment = Experiment.create(experiment_params)
+    redirect_to experiment_path @experiment
+  end
+  def update
+    @experiment = Experiment.find(params[:id])
+    @experiment.update_attributes!(experiment_params)
+    flash[:notice] = 'Changes has been saved'
+    redirect_to experiment_path @experiment
+  end
+  def destroy
```
Comment (jrhorn424) [inline app/controllers/experiments_controller.rb:4]: InheritedResources. The goal of Inherited Resources is to take care of the standardized CRUD actions. If you follow the same manual pattern every has used in thousands of restful controllers, you won't need to write them when your controller inherits from Inherited Resources.

File: app/controllers/experiments_controller.rb [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+  end
+  def show
+    @experiment = Experiment.find(params[:id])
+    @sessions = @experiment.sessions
+  end
+  def create
+    @experiment = Experiment.create(experiment_params)
+    redirect_to experiment_path @experiment
+  end
+  def update
```
Comment (jrhorn424) [inline app/controllers/experiments_controller.rb]: Same.

File: app/controllers/experiments_controller.rb [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    @experiments = Experiment.all
+  end
+  def new
+    @experiment = Experiment.new
+  end
+  def show
+    @experiment = Experiment.find(params[:id])
+    @sessions = @experiment.sessions
+  end
+  def create
```
Comment (jrhorn424) [inline app/controllers/experiments_controller.rb]: Same.

File: app/controllers/experiments_controller.rb [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    @experiment = Experiment.create(experiment_params)
+    redirect_to experiment_path @experiment
+  end
+  def update
+    @experiment = Experiment.find(params[:id])
+    @experiment.update_attributes!(experiment_params)
+    flash[:notice] = 'Changes has been saved'
```
Comment (jrhorn424) [inline app/controllers/experiments_controller.rb]: Flashes are already provided.

File: app/controllers/profiles_controller.rb [hunk 15]
```diff
@@ -1,30 +1,46 @@
 class ProfilesController < InheritedResources::Base
-  # before_filter :authenticate_user!
-  defaults singleton: true
```
Comment (jrhorn424) [inline app/controllers/profiles_controller.rb:3]: Needed by InheritedResources.

File: app/controllers/profiles_controller.rb [hunk 16]
```diff
@@ -1,30 +1,46 @@
...
-  belongs_to :user
```
Comment (jrhorn424) [inline app/controllers/profiles_controller.rb:4]: Likewise, I believe.

File: app/controllers/profiles_controller.rb [hunk 17]
```diff
@@ -1,30 +1,46 @@
...
+  actions :show, :update, :create
 
   def show
-    redirect_to new_user_profile_path if current_user.profile.blank?
-    show! unless current_user.profile.blank?
+    if(current_user.profile.nil?)
+      @profile = Profile.new
+    else
+      @profile = current_user.profile
+    end
```
Comment (jrhorn424) [inline app/controllers/profiles_controller.rb:9]: This change is not necessary. Please revert. InheritedResources takes care of all this. If this is related to viewing of other users' profiles. Let's work through this together.

File: app/controllers/profiles_controller.rb [hunk 18]
```diff
@@ -1,30 +1,46 @@
...
+  end
+  def update
```
Comment (jrhorn424) [inline app/controllers/profiles_controller.rb:11]: IR

File: app/controllers/profiles_controller.rb [hunk 19]
```diff
@@ -1,30 +1,46 @@
...
+    @profile = current_user.profile
+    @profile.update(profile_params)
+    @profile.save
+    flash[:notice] = 'Profile has been saved'
+    redirect_to :back
+  end
+  def create
```
Comment (jrhorn424) [inline app/controllers/profiles_controller.rb:19]: IR

File: app/controllers/profiles_controller.rb [hunk 20]
```diff
@@ -1,30 +1,46 @@
...
+    @profile = Profile.new(profile_params)
+    @profile.update(user_id: current_user.id)
+    @profile.save
+    flash[:notice] = 'Profile has been saved'
+    redirect_to :back
   end
 
   private
-    def permitted_params
```
Comment (jrhorn424) [inline app/controllers/profiles_controller.rb:13]: This is the IR pattern for strong parameters.

File: app/controllers/sessions_controller.rb [hunk 21]
```diff
@@ -0,0 +1,50 @@
+class SessionsController < InheritedResources::Base
```
Comment (jrhorn424) [inline app/controllers/sessions_controller.rb:1]: For this entire change, apply everything I've said about IR.

File: app/models/experiment.rb [hunk 22]
```diff
@@ -1,5 +1,7 @@
 class Experiment < ActiveRecord::Base
   has_paper_trail
-
+  has_many :sessions
+  has_many :assignments
+  has_many :subjects, through: :assignments
```
Comment (jrhorn424) [inline app/models/experiment.rb:5]: Looks good.

File: app/models/session.rb [hunk 23]
```diff
@@ -1,4 +1,7 @@
 class Session < ActiveRecord::Base
   has_paper_trail
+  belongs_to :experiment
   belongs_to :lab, inverse_of: :sessions
+  has_many :registrations
+  has_many :subjects, through: :registrations
```
Comment (jrhorn424) [inline app/models/session.rb:6]: Looks good.

File: app/models/user.rb [hunk 24]
```diff
@@ -14,14 +14,14 @@ def name
     "#{self.profile.first_name} #{self.profile.last_name}"
   end
 
-  %i[administrator experimenter subject].each do |method|
-    define_method "is_#{method}?" do
-      self.type == method.to_s.capitalize
-    end
```
Comment (jrhorn424) [inline app/models/user.rb:20]: What was wrong with this code?

File: app/models/user.rb [hunk 25]
```diff
@@ -14,14 +14,14 @@ def name
...
+    self.type == "Administrator" or self.type == "Experimenter"
   end
-
-  def is_admin?
-    false
+  def is_administrator?
+    self.type == "Administrator"
+  end
+  def is_subject
+    self.type == "Subject"
```
Comment (jrhorn424) [inline app/models/user.rb:24]: Not necessary. That's what all that magic above was for.

File: app/views/devise/sessions/new.html.slim [hunk 26]
```diff
@@ -1,6 +1,6 @@
 div.row
   h2 Sign in
-  = form_for(resource, :as => resource_name, :url => session_path(resource_name)) do |f|
```
Comment (jrhorn424) [inline app/views/devise/sessions/new.html.slim:3]: This is for IR. Please revert.

File: app/views/experiments/_sessions.html.slim [hunk 27]
```diff
@@ -0,0 +1,19 @@
...
+    tbody
+      - @sessions.each do | session |
+        tr id="session_#{session.id}"
+          td data-href = "#{session_path(session)}"
+            = l session.start_time.to_date
+          td data-href = "#{session_path(session)}"
+            = session.subjects.count
+          td
+            .btn-group.pull-right
+              = link_to 'Remove', session, remote: true, method: :delete, :class => 'btn btn-xs btn-danger remove-link', data: { confirm: 'Are you sure?' }
```
Comment (jrhorn424) [inline app/views/experiments/_sessions.html.slim]: LGTM

File: app/views/experiments/index.html.slim [hunk 28]
```diff
@@ -0,0 +1,30 @@
...
+        tr id="experiment_#{experiment.id}"
+          td data-href = "#{experiment_path(experiment)}"
+            = experiment.name
+          td data-href = "#{experiment_path(experiment)}"
+            = experiment.type
+          td data-href = "#{experiment_path(experiment)}"
+            = experiment.subjects.count
+          td
+            .btn-group.pull-right
+              = link_to 'Remove', experiment.becomes(Experiment), remote: true, method: :delete, :class => 'btn btn-xs btn-danger remove-link', data: { confirm: 'Are you sure?' }
```
Comment (jrhorn424) [inline app/views/experiments/index.html.slim:30]: LGTM

File: app/views/experiments/new.html.slim [hunk 29]
```diff
@@ -0,0 +1,33 @@
...
+    | experiment
+= form_for(@experiment, :html => { :class => "form-horizontal" }) do |f|
+  - if @experiment.errors.any?
+    .alert.alert-danger.alert-dismissable
+      button.close type="button" data-dismiss="alert" aria-hidden="true"
+        | &times;
+      h4= "#{pluralize(@experiment.errors.count,"error")} prohibited this experiment from being saved:"
+      ul
+        - @experiment.errors.full_messages.each do |msg|
+          li= msg
```
Comment (jrhorn424) [inline app/views/experiments/new.html.slim:13]: I have much of this defined as a partial somewhere, and I believe IR was providing support for it. You can look at the profile form for the signature to pull in errors (should be one line).

File: app/models/user.rb (hunk 24 above)
Comment (ppodolsky) [inline app/models/user.rb:20]: Plz, look at is_experimenter?. That was wrong.

File: app/models/user.rb (hunk 25 above)
Comment (ppodolsky) [inline app/models/user.rb:24]: Same

File: app/views/pages/help.html.slim [hunk 30]
```diff
@@ -1,8 +1,6 @@
 h2 Help
 ul.list-group
   - if controller_name != 'sessions'
```
Comment (jrhorn424) [inline app/views/pages/help.html.slim:3]: This shouldn't be necessary if I understand correctly. This modal should never appear if you're already signed in, and you shouldn't be able to see sessions if you aren't signed in.

File: app/views/profiles/_form.html.slim [hunk 31]
```diff
@@ -1,4 +1,4 @@
-= form_for(resource, url: (params[:action] == :edit ? edit_user_profile_path : user_profile_path)) do |f|
```
Comment (jrhorn424) [inline app/views/profiles/_form.html.slim:1]: Why was this changed?

File: app/views/profiles/_form.html.slim [hunk 32]
```diff
@@ -19,22 +19,19 @@
       div.form-group
         = f.label :gender, class: 'control-label col-sm-2'
         div.col-sm-8
-          = f.select :gender,
-            options_for_select(['M','F'], resource.gender),
```
Comment (jrhorn424) [inline app/views/profiles/_form.html.slim:23]: Again, resource is something added by IR.

File: app/views/profiles/show.html.slim [hunk 33]
```diff
@@ -1,11 +1,2 @@
...
-    |  (#{link_to 'edit', edit_user_profile_path unless @profile.blank?})
-- if @profile.blank?
-  div.alert.alert-danger
-    p
-      | Your profile is blank! Please #{link_to('create one', new_user_profile_path)}.
-      |  You will be ineligible to receive invitations until it is complete.
-.row
-= render 'profile' unless @profile.blank?
+h2 Update your profile
+= render 'form'
```
Comment (jrhorn424) [inline app/views/profiles/show.html.slim:2]: Is the profile now editable inline?

File: config/routes.rb [hunk 34]
```diff
@@ -6,9 +7,11 @@
   # user accounts and profiles
   devise_for :users
 
-  resources :users, only: :none do
-    resource :profile
```
Comment (jrhorn424) [inline config/routes.rb:10]: Blowing this away pollutes the output of `rake routes -T` with unnecessary paths from Devise.

File: config/routes.rb [hunk 35]
```diff
@@ -6,9 +7,11 @@
...
+  resources :sessions
+  resources :experiments do
+    resources :sessions
   end
+  resource :profile
```
Comment (jrhorn424) [inline config/routes.rb:14]: I have a mild preference for this to be namespaced under the user.

File: app/models/user.rb (hunk 25 above)
Comment (jrhorn424) [inline app/models/user.rb:24]: Yes, but you don't need to redefine it. The method is already defined. And you forgot the `?` at the end.

File: app/models/user.rb (hunk 24 above)
Comment (jrhorn424) [inline app/models/user.rb:20]: I see, thank you for the context. If this is the only override necessary, you don't need to redefine all the methods. If for some reason there is a collision, you could always `undef` the method you need to override, but I don't think it's necessary.

File: Gemfile (hunk 2 above)
Comment (jrhorn424) [inline Gemfile]: OK, just wanted to let you know they do the same thing. They're both authorization gems. Shouldn't need both. I hadn't added any code supporting authority beyond installing an initializer (I'm not even sure that was done).

File: app/assets/javascripts/experiments.js (hunk 4 above)
Comment (jrhorn424) [inline app/assets/javascripts/experiments.js:8]: I may not be understanding you, but this is how I've seen it done in the past.

``` ruby
  <% for post in @posts %>
 ...
      <td><%= link_to "Destroy", post, :confirm => 'Are you sure?', :method => :delete %></td>
    </tr>
  <% end %>
```

Does that fail to remove the row? It should just work, but I'm not sure if anything changed between rails 3 and 4.

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (ppodolsky) [inline app/controllers/application_controller.rb:14]: This question should be reviewed in the context of changed ProfileController. Let's discuss it there.

Comment (jrhorn424) [inline app/controllers/application_controller.rb:14]: OK, thanks.

File: app/controllers/experiments_controller.rb (hunk 5 above)
Comment (ppodolsky) [inline app/controllers/experiments_controller.rb]: Okay, I will change it. The reason is that methods will appear again in several commits, because the same controller is used to deal with viewing Experiments for admins and for subjects. So, filtering is going on

File: app/controllers/experiments_controller.rb (hunk 7 above)
Comment (ppodolsky) [inline app/controllers/experiments_controller.rb]: I will fix

File: app/controllers/experiments_controller.rb (hunk 8 above)
Comment (ppodolsky) [inline app/controllers/experiments_controller.rb]: same

File: app/controllers/experiments_controller.rb (hunk 7 above)
Comment (jrhorn424) [inline app/controllers/experiments_controller.rb]: Understood, and thanks! If you don't want to use IR, I'm OK with that. I just want the controllers to be consistent. So whatever you prefer.

As for all the notes: I didn't want anything to get lost. Don't feel obliged to respond to every comment.

File: app/views/devise/sessions/new.html.slim (hunk 26 above)
Comment (ppodolsky) [inline app/views/devise/sessions/new.html.slim:3]: Right note. But! There was some trouble - naming conflict. session_path from session resource against session_path from devise. Some gang bang occurred while I was debugging it.

Comment (jrhorn424) [inline app/views/devise/sessions/new.html.slim:3]: > Some gang bang occurred while I was debugging it.

That just made my day. :smile:

I didn't even think of that! It might be better to rename session resources to `ExperimentSession`. Does that conflict with any joins?

File: app/views/profiles/show.html.slim (hunk 33 above)
Comment (ppodolsky) [inline app/views/profiles/show.html.slim:2]: Yes, now it is editable by default. I believe from my experience that people are going for their profile pages only for changing something, not to look at it. Okay, may be there are some boozy students who will try to remember who are they via looking their profile, but it is not our target audience

File: app/assets/javascripts/experiments.js (hunk 4 above)
Comment (ppodolsky) [inline app/assets/javascripts/experiments.js:8]: Yes, this code destroy row on the server-side. The row is erasing from database, all right. But the interface row (tr) in our html code won't disappear until we will reload the page. I guess, it can frustrate user a lot, if the interface won't react on his actions:)

File: app/controllers/experiments_controller.rb (hunk 14 above)
Comment (ppodolsky) [inline app/controllers/experiments_controller.rb]: Sorry, I've missed it. Does it mean that every action in InheritedResource controller will call appropriate flash notice?

File: app/views/devise/sessions/new.html.slim (hunk 26 above)
Comment (jrhorn424) [inline app/views/devise/sessions/new.html.slim:3]: Ah, it just occurred to me. Let's just namespace the _routes_, something like:

``` ruby
resources :experiments do
  resources :sessions
end
```

File: app/views/profiles/show.html.slim (hunk 33 above)
Comment (jrhorn424) [inline app/views/profiles/show.html.slim:2]: Agreed. I like this change. Thanks!

File: app/assets/javascripts/experiments.js (hunk 4 above)
Comment (jrhorn424) [inline app/assets/javascripts/experiments.js:8]: I see, thanks!

File: app/controllers/profiles_controller.rb (hunk 17 above)
Comment (ppodolsky) [inline app/controllers/profiles_controller.rb:9]: Thanks! I understood your wishes, we should consider most of your notes together. The main issue is that there is a singleton mechanics in IR, but I recreated it by hands. All others corrections are consequences of this.
Actually, this change was necessary, because I moved away from the concept of IR. So, I will overview what is IR and commit corrections tomorrow.

File: app/controllers/experiments_controller.rb (hunk 14 above)
Comment (jrhorn424) [inline app/controllers/experiments_controller.rb]: Yes, I think so. If you want to change the message, you can do so in I18n.

From the docs:

> Using responders will set the flash message to :notice and :alert. You
> can change that through the following configuration value:
> 
> InheritedResources.flash_keys = [ :success, :failure ] Notice the
> CollectionResponder won't work with InheritedResources, as
> InheritedResources hardcodes the redirect path based on the current scope
> (like belongs to, polymorphic associations, etc).

I believe I've already set the flash keys to be appropriate for bootstrap.

Speaking of which, you seem to be working with bootstrap pretty well. I had planned on removing it, but I'll leave it in for now. Have you ever worked on cleaning up the views by using sass/less mixins?

File: app/views/devise/sessions/new.html.slim (hunk 26 above)
Comment (ppodolsky) [inline app/views/devise/sessions/new.html.slim:3]: Hope, we won't have to do it. I'm looking IR docs now, and my sight has fallen onto the string:
defaults :route_prefix => 'blablabla'
I believe that what we need.

File: app/controllers/profiles_controller.rb (hunk 16 above)
Comment (ppodolsky) [inline app/controllers/profiles_controller.rb:4]: It is needed in the case when route is something like /users/profile
When it is /profile, it is unnecessary.
We should choose which route should be used. Let's consider pros and cons, you are welcome.
/profile pros
1. It is plain. We have a requirement that only subject can have profile, /users/profile can mean that every user can have profile.
2. /profile means "my profile". I can send this url to person and say - hey, dude, that what you need to fill in. Don't need to find his id to compose correct url.
3. We can divide logic for admins /users/{id}/profile urls (which are needed to see information about victims) and subjects (which are allowed to edit only their own profiles) /profile url by natural way, without dancing inside controller for checking if we can modify this and cannot modify that. But, of course, this pros is arguable,  I dunno opportunities of IR to protect resources.

**Comment by jrhorn424** [comment]
It's a good idea to put it somewhere. I'm not sure it will autolink if you put it in the title. If you say something like

```
Closes #20
```

in the description of a pull request, it will automatically close that ticket when the pull request is merged. Works with multiple tickets.

**Comment by jrhorn424** [comment]
Great work, @PaulPodolsky
